File: solutions/java/perfect-numbers/1/src/main/java/NaturalNumber.java
class NaturalNumber {
    final int number;
    final int aliquot;
    
    NaturalNumber(int number) {
        if (number <= 0) {
            throw new IllegalArgumentException("You must supply a natural number (positive integer)");
        }
        int aliquot = 0;
        for (int i = 1; i <= number / 2; ++i) {
            if (number % i == 0) {
                aliquot += i;
            }
        }
        this.number = number;
        this.aliquot = aliquot;
    }

    Classification getClassification() {
        if (aliquot == number) {
            return Classification.PERFECT;
        } else if (aliquot > number) {
            return Classification.ABUNDANT;
        } else {
            return Classification.DEFICIENT;
        }
    }
}

File: solutions/kotlin/perfect-numbers/1/src/main/kotlin/NaturalNumber.kt
enum class Classification {
    DEFICIENT, PERFECT, ABUNDANT
}

fun classify(naturalNumber: Int): Classification {
    require(naturalNumber > 0) {
        "Must provide a positive number"
    }
    return (1..<naturalNumber).filter{ divisor ->
        naturalNumber % divisor == 0
    }.sum().let{ aliquotSum ->
        when {
            aliquotSum == naturalNumber -> Classification.PERFECT
            aliquotSum < naturalNumber -> Classification.DEFICIENT
            else -> Classification.ABUNDANT
        }   
    }
}